src/mbgl/renderer/render_layer.cpp

```diff
@@ -71,8 +76,60 @@ bool RenderLayer::needsRendering(float zoom) const {
            && baseImpl->maxZoom >= zoom;
 }
 
-void RenderLayer::setRenderTiles(std::vector<std::reference_wrapper<RenderTile>> tiles) {
-    renderTiles = std::move(tiles);
+void RenderLayer::setRenderTiles(RenderTiles tiles, const TransformState& state) {
+    renderTiles = filterRenderTiles(std::move(tiles));
+    sortRenderTiles(state);
+}
+
+RenderLayer::RenderTiles RenderLayer::filterRenderTiles(RenderTiles tiles) const {
+    return filterRenderTiles(std::move(tiles), [](const auto& tile){ return !tile.tile.isRenderable() && tile.tile.holdForFade(); });
+}
+
+void RenderLayer::sortRenderTiles(const TransformState&) {
+    std::sort(renderTiles.begin(), renderTiles.end(), [](const auto& a, const auto& b) { return a.get().id < b.get().id; });
+}
+
+// We only need clipping when we're drawing fill or line layers.
+bool RenderLayer::needsClipping() const {
+    return false;
+}
+
+bool RenderLayer::hasSymbols() const {
+    return false;
+}
+
+void RenderLayer::update() {
+    // no-op
+}
+
+optional<Color> RenderLayer::getSolidBackground() const {
+    return nullopt;
+}
+
+RenderLayer::RenderTiles RenderLayer::filterRenderTiles(RenderTiles tiles, const std::function<bool (RenderTile&)>& filter) const {
+    RenderTiles filtered;
+
+    for (auto& tileRef : tiles) {
+        auto& tile = tileRef.get();
+        if (filter(tile)) {
+            continue;
+        }
+
+        if (tile.tile.getBucket(*baseImpl)) {
+            tile.used = true;
+            tile.needsClipping = needsClipping();
```

Comment:
One tile can have layers of many types. If any non-fill/line layer come after a fill/line layer, they'll reset `needsClipping` to false here, and break rendering for the fill/line layers.

Reply:
Done